src/sources/udp.rs

```diff
@@ -0,0 +1,220 @@
+impl SourceConfig for UdpConfig {
+    fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> Result<super::Source, String> {
+        let host_key = self.host_key.clone().unwrap_or(event::HOST.clone());
+        Ok(udp(self.address, host_key, out))
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+}
+
+pub fn udp(address: SocketAddr, host_key: Atom, out: mpsc::Sender<Event>) -> super::Source {
+    let out = out.sink_map_err(|e| error!("error sending event: {:?}", e));
+
+    Box::new(
+        future::lazy(move || {
+            let socket = UdpSocket::bind(&address).expect("failed to bind to udp listener socket");
+
+            info!(message = "listening.", addr = &field::display(address),);
+
+            future::ok(socket)
+        })
+        .and_then(move |socket| {
+            let host_key = host_key.clone();
+            let mut new_line_decoder = BytesDelimitedCodec::new(b'\n');
+            UdpFramed::new(socket, BytesCodec::new())
+                .map(move |(mut bytes, addr)| {
+                    // A packet has come.
+                    // UDP processes messages per packet, where messages are separated by newline.
+                    let mut events = Vec::new();
+                    while let Some(Some(line)) = new_line_decoder
+                        .decode_eof(&mut bytes)
+                        .map_err(|e| error!("{}", e))
+                        .ok()
+                    {
+                        // One message
+                        let mut event = Event::from(line);
+
+                        event
+                            .as_mut_log()
+                            .insert_implicit(host_key.clone().into(), addr.to_string().into());
+
+                        trace!(
+                            message = "Received one event.",
+                            event = field::debug(&event)
+                        );
+
+                        events.push(event);
+                    }
+                    iter_ok(events)
+                })
+                // Flatten messages from single packet
+                .flatten()
+                .map_err(|e: io::Error| error!("error reading datagram: {:?}", e))
+                .forward(out)
+                .map(|_| info!("finished sending"))
+        }),
+    )
+}
+
+#[cfg(test)]
+mod test {
+    use super::UdpConfig;
+    use crate::event;
+    use crate::test_util::next_addr;
+    use crate::topology::config::{GlobalOptions, SourceConfig};
+    use futures::{future::poll_fn, sync::mpsc, try_ready, Future, Stream};
+    use std::{
+        mem,
+        net::{SocketAddr, UdpSocket},
+        thread,
+        time::Duration,
+    };
+    use tokio::prelude::Async;
+
+    fn send_lines<'a>(addr: SocketAddr, lines: impl IntoIterator<Item = &'a str>) -> SocketAddr {
+        let bind = next_addr();
+
+        let socket = UdpSocket::bind(bind)
```

Comment:
would we want to use tokio::net::UdpSocket here? so that we can spawn it onto the runtime instead of blocking the thread?

Reply:
We definitely have some tests that use synchronous IO and I think that's totally fine. We should be more concerned with keeping the tests simple and easy to write than having them run perfectly asynchronously (we usually `block_on` anyway).